actionview/lib/action_view/helpers/tag_helper.rb

```diff
@@ -72,10 +73,50 @@ def initialize(view_context)
         #   <input <%= tag.attributes({ type: :text }, { id: "search" }, { aria: { label: "Search" } }, aria: { disabled: true }) %> >
         #   # => <input type="text" id="search" aria-label="Search" aria-disabled="true">
         #
-        ruby2_keywords def attributes(*attributes)
-          attributes = attributes.tap(&:flatten!).tap(&:compact_blank!)
+        # When called outside of a rendering context, <tt>tag.attributes</tt>
+        # will return a <tt>Hash</tt>-like object that knows how to render
+        # itself to HTML:
+        #
+        #   primary = { class: "bg-red-500 text-white" }
+        #   large = { class: "text-lg p-4" }
+        #
+        #   button_tag "Click me!", tag.attributes(primary, large)
+        #   # => <button name="button" type="submit" class="bg-red-500 text-white text-lg p-4">Click me!</button>
+        #
+        #   tag.button "Click me!", id: "cta", **tag.attributes(primary, large)
+        #   # => <button id="cta" class="bg-red-500 text-white text-lg p-4">Click me!</button>
+        #
+        # === Token list support
+        #
+        # Attribute merging will account for token lists attributes (like <tt>class</tt> and <tt>aria-labelledby</tt>) by combining values from left to right
+        #
+        #   <input <%= tag.attributes({ class: "font-bold" }, { class: ["text-sm", "text-gray-700"] }, class: "p-2") %> >
+        #   # => <input class="font-bold text-sm text-gray-700 p-2" >
+        #
+        # To override token list merging for an attribute, pass its name with a `!` suffix:
+        #
+        #   attributes = tag.attributes({ class: "default" }, { class!: "first-override" }, class!: "second-override")
+        #   attributes.to_h # => { class: "second-override" }
+        #   attributes.to_s # => "class=\"second-override\""
+        #
+        # Keys that end with <tt>!</tt> are only temporary. Accessing
+        # <tt>attributes[:class!]</tt> will return <tt>nil</tt> in the code sample
+        # above.
+        #
+        # === Configuring token list support
+        #
+        # To treat addition attributes as token lists, add values to the <tt>config.action_view.token_lists</tt> value:
+        #
+        #   config.action_view.token_lists << "data-action"
+        #   config.action_view.token_lists << "data-controller"
+        #   config.action_view.token_lists << /data-(.*)-target/
+        #
+        ruby2_keywords def attributes(*hashes)
+          attributes = ActionView::Attributes.new(@view_context) do |value|
+            tag_options(value).to_s.strip.html_safe
```

Comment:
This style of serialization is an attempt to keep this changeset conservative, and avoid moving the private `TagBuilder#tag_options` implementation out of that class and into `ActionView::Attributes`. If `ActionView::Attributes` is deemed viable, that work could be done in a future PR.